jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -2307,37 +2307,21 @@ public boolean equals(Object object) {
 		return false;
 	}
 
-	private static final void deduplicateStrings(String s1, String s2) {
-		if ((s1.count ^ s2.count) >= 0) {
+	private static final void deduplicateStrings(String s1, Object value1, String s2, Object value2) {
+		if (!enableCompression || null == compressionFlag || (s1.count ^ s2.count) >= 0) {
```

Comment:
Is this meant to be checking whether the string is compressed and only dedup'ing non-compressed strings?

The check for compression in `isCompressed()` is:
```
if (enableCompression && (null == compressionFlag || count >= 0)) {
```
which should be this negated:
```
if (!enableCompression || (null != compressionFlag && count < 0)) {
```

Did I miss something?

Reply:
This test ensures that we only deduplicate strings that are both compressed or both uncompressed.

When one string is compressed and the other isn't, we can't deduplicate because doing so would require updating both `value` and `count`, and other threads could see an inconsistent state.

If `!enableCompression`, then both strings are always uncompressed. If OTOH `enableCompression` but `null == compressionFlag`, then both strings must be compressed. So it's only necessary to check the compression bits when `enableCompression && null != compressionFlag`.